### ISSUE TITLE
Save UI state periodically to prevent loss on shutdown

### ISSUE THREAD

**Issue by jankalfus** [microsoft/vscode#12058]
- VSCode Version: 1.5.2
- OS Version: Windows 10

Steps to Reproduce:
1. Let VS Code open, lock the computer
2. Let Windows install updates and automatically restart the computer

After logging into Windows and launching VS Code, VS Code opens the project that was opened last time, but it doesn't restore editors to the state they were in last time. It restores some much older state instead (probably from last time the "Close folder" menu option was used). Very annoying.

**Comment by bpasero** [comment]
@jankalfus this indicates that Windows is not shutting down VS Code gracefully but rather terminates it. We would need to periodically save UI state to prevent that.

**Comment by jankalfus** [comment]
I have also noticed that when using F12 (Go to Definition), VS Code can be confused with certain files (probably those that weren't existing in the restored state, but were created later).

The issue is the following (BTW this relates to TypeScript .tsx and .ts files, I haven't tested any others):
When using F12 with a symbol, VS Code opens a new editor tab with an identical file opened and sets the cursor at the import, instead of going to the symbol import in the currenty opened editor tab. Expected behavior is to just set the cursor in the currently opened tab.

**Comment by YMIndustries** [comment]
I'm not sure that periodically saving the UI state is the only way to achieve this. With other Windows programs, they appear to be able to execute some code when the computer shuts down. For instance, Paint.NET will prompt you to save any unsaved changes, and will even halt the shutdown until you have given a response.

I'm having trouble finding up-to-date documentation about this, but it seems in Vista Windows was changed to allow programs 2 seconds to exit once a shutdown is initiated (https://blogs.msdn.microsoft.com/oldnewthing/20070416-00/?p=27243/), and I can't see any sign that this has been changed since. Would it be possible to use this time to store the UI state?

**Comment by petermilan** [comment]
+1

**Comment by bpasero** [comment]
From https://github.com/Microsoft/vscode/issues/62844: this also applies to UI state we accumulate on the main side such as the list of opened windows and window dimensions.

**Comment by Ruslan0Dev** [comment]
Where solution? My vscode restore older state instead (After every power failure or critical shutdown). Whereas sublimetext never failed. Please borrow fault tolerance from sublimetext. Thanks.

**Comment by Daijobou** [comment]
Today windows make a update and vscode again lost all actual files and show instead (tabs)files that long ago deleted or moved. So now I must search again ~50 files for my project i'm working with. :(

I would wish vscode not only store the actual files before restart windows, its prevent restart when files are not saved at all.

Offtopic: I hate windows 10 for his forced restarts. I never shutdown or restart my system. I use hibernate mode, because I have many tools and files open. I hope the next windows is more modern and will be more linux, so restarts are not required. Its feel like year 1995.

**Comment by atiqorin** [comment]
Regarding closed: #65908. There is something else going on. I did update vscode yesterday. Had atleast 13 tabs pinned and open. 7-8 of them were opened automatically after I started my pc yesterday morning. So while going out of work, as usual I didn't close vscode and shut down my pc. I did restarted vscode two times yesterday. Once after the vscode update and once after I had a corruption on my yarn cache. Few files were probably not opened because before opening vscode, I stashed my changes, merge another branch. Maybe the files that were newly added on my branch were removed while checking out other branch caused the files to not auto open in vscode (Just speculating, I didn't even open vscode while doing the merge or stashing. Opened vscode after I checked out my branch and popping my stash. So the files were there at the time of vscode opening.). Now long story short, when I opened vscode today, just 3 files and an untitled tab were open. One I keep open all the time. Never had issue with that file. One other I am sure I opened yesterday by  control clicking the interface usage. Another one was open for 3-4 days. However all the other opened tabs are closed, even the ones that opened yesterday automatically when I opened pc.

I don't know how to reproduce this. I believe this is related to this issue and creating another issue will just cause duplication, so commenting here. Thanks.

**Comment by youurayy** [comment]
tl;dr this thing is flat broken

**Comment by agilgur5** [comment]
I'm having this problem on my Mac too, whenever VSCode freezes up or I have to restart my computer, I re-open and I'm back to a state from a month ago :/

I agree with @youurayy that this feature shouldn't rely on or require a graceful shutdown, one can't guarantee that will ever happen. Makes sense to save the state on-change to prevent this as other tools do

**Comment by ddfridley** [comment]
I agree with Somebi's suggestion.  I'll also add that my PC was off, when windows 10 did the update. I don't know if that effects what events VS code gets (if any).   I'll also add that I wish the state information was stored in each folder that was open (I have 3 separate  instances in their own folders open), and then when VS code starts up, it would open up 3 instances again.  I understand this is not the most urgent features, and I really appreciate VS Code.   Thank you for the work you do.

**Comment by Noiredd** [comment]
Any attention to this? It's been over 2 years since this was raised.

Below, alternative steps to reproduce the issue, independent of your OS:
1. Launch an existing session of Code, say there are files A and B open
2. Close tab B
3. Open an existing file C but don't make any modifications to it
4. Create a new tab (let's call it D) and write "foo" in it, without saving
5. Kill the code process (or hard-reset your machine)
6. Launch Code again

As a result, you will see tabs A, B and D open. B will be reopened, while C will be lost.
If you created a new window and only used it to view existing files (more of the C case), this will mean the loss of an entire window. If you had multiple windows and closed one of them before step 5, it will be revived.

**Comment by Yanpas** [comment]
Steps to Reproduce:

1. Open vscode
2. Open some files, wait some time
3. killall -9 code
4. Open vscode again, these files are not opened

I never close vscode on my work machine, sometimes power goes down, PC gets rebooted and I have to manually open each file that was opened after vscode launch. Chrome saves opened tab periodically, add this to vscode too.

**Comment by adbakke** [comment]
What's the lower-limit of comments on a bug before it gets addressed for this project?

🤔

**Comment by mar-ses** [comment]
Just commenting to also note that this is an issue on Ubuntu with shutdowns (see #36964), although I assume the proposed fix would also work for all systems.

**Comment by ackvf** [comment]
#74083

![vscode-crash-tabs](https://user-images.githubusercontent.com/2313018/58270550-5d8d2200-7d8a-11e9-90e6-f1518eb574a3.gif)

**Comment by youurayy** [comment]
Could someone get this one moving? 
@alexr00 @joaomoreno @Tyriar

One of the most popular editors (if not No.1), and not saving window state across process restarts, in 2019...

**Comment by Tyriar** [comment]
@youurayy please don't ping random people, we get way too many notifications already.

**Comment by youurayy** [comment]
@Tyriar sorry, I've tried to reach someone currently active on bug triage, as this one seems to be wrongly categorized as a `feature-request` when it's in fact a bug, as I think this is commonly expected behavior. (Or perhaps the README should say "Doesn't persist window state across restarts.")

**Comment by Meai1** [comment]
This is pretty insane, it's an issue EVERY SINGLE DAY. It's the most visible, jarring problem that practically everyone is going to have. Just today I had yet more long deleted files show up in my workspace. They keep showing up, no matter how many times I remove them. The entire tab/open files experience is totally broken. For years(!).

**Comment by Glamurio** [comment]
Just today I was wondering if anyone else has been running into this issue, and it turns out I'm not alone. Pretty please get this fixed, it shouldn't be a surprise that this is really irritating when working with VSC.

**Comment by adbakke** [comment]
Can anyone explain why this is tagged as a feature-request and not a bug? If the behavior of a normal shutdown is to save workbench state and reopen in that state later, why would shut down occurring through the operating system not be exactly the same? If it should be then this would be a bug, not a feature request. 

I am open to understanding why it is a feature request and not a bug.

**Comment by felixpenrose** [comment]
+1 this is very jarring - a saved editor state is the easiest way to continue what I was working on the previous day. Just switched from sublime and this is so far the most annoying issue in VSC

**Comment by VigiFyre** [comment]
Title and tag changes make me worry that the issue might have been misunderstood.

It's not that VSC doesn't keep backups for cases like sudden power loss (by the way, it absolutely should). It's that VSC doesn't save the session even when Windows is shut down gracefully. When Windows is shutting down, apps are even able to ask Windows to wait up until app was able to save everything, so i'm pretty sure that Windows doesn't just kill apps when it's shutting down or restarting. And it doesn't happen only when Windows decides to update, for example – it happens every time PC is shut down for the night.

It's not a feature request, it's a bug.

Using VSC to work feels a little bit like walking on a minefield:
Forgot to manually open VSC from task bar before double-clicking any text file? Whoops, your entire session is rewritten.
Forgot to manually close VSC before shutting down the PC? Whoops, your session is gone again.

**Comment by akontsevich** [comment]
> It's not that VSC doesn't keep backups for cases like sudden power loss (by the way, it absolutely should). It's that VSC doesn't save the session even when Windows is shut down gracefully. 

It happen not only in Windows, but in Linux too.

**Comment by RafaelQuirino** [comment]
I saw that people (and even the title of the issue) are suggesting that it is necessary to periodically save the state of the app. But couldn't we just save the state of open files and folders each time a file or folder is opened ?

**Comment by youurayy** [comment]
@RafaelQuirino  no I think the consensus here is to save-on-change (i.e. https://github.com/microsoft/vscode/issues/12058#issuecomment-412556800). But you are right, the issue title should be adapted accordingly.

**Comment by wnsnty** [comment]
This issue has been referenced more than 20 times and has been continually commented on over the past 3 years. At what point does an issue get reevaluated? Because the tags on this one are wrong.

In what world is "UI state is lost after program exited abruptly" considered "working as intended"?

**Comment by RafaelQuirino** [comment]
I opened an issue that was closed as a duplicate of this one. I noticed that this one, despite being kinda old, had a lot of activity recently. A lot of people is annoyed by this. Are you planning to solve it or not ? Because it is very serious to me, to the point i have to abandon the application. Someone on the team, please, solve this problem.

**Comment by wmertens** [comment]
Curious, I use VSCode on macOS and I never noticed this problem, and I've had my fair share of crashes. Is it limited to Windows? What is the cause?

**Comment by akontsevich** [comment]
@wmertens no, as You can see above it happens in Linux as well.

**Comment by wmertens** [comment]
@akontsevich I wonder what macOS is doing differently, or maybe I was lucky enough not to encounter it? Is there a way to reproduce it 100%? Maybe some editor setup steps followed by a `kill -9`?

**Comment by petermilan** [comment]
Yeah happens on linux as well (In my case it is Kubuntu). It is really annoying to see same opened files every single morning. And trying to recall what files were opened the day before.

**Comment by petermilan** [comment]
I am seriously thinking about a workarround: making custom command for  shutdown of VS code like: 

```
#!/bin/sh

$(kill $(ps -ef |grep `whoami`| grep -m1 "/usr/share/code/code --unity-launch" | cut -c10-15))

```
The command above will stop VS code and remember last state of it.
The grep search in process `/usr/share/code/code --unity-launch` can be different on different systems.

Wondering if I could call it with `shutdown -h now` and use it as custom shutdown script,
But will need some time to test or tune it,  or even perhaps somehow call as a script before shutdown of OS.

**Comment by wmertens** [comment]
How is that different from shutdown? All processes get a SIGTERM.

**Comment by felixpenrose** [comment]
How many years does this bug have to exist before it gets addressed? (and yes, it's a bug, not a feature-request). Every morning I start up my PC, yet again the sessions starts in a random saved state from days ago, most of the time it's not even the same work space, meaning I've lost my checkpoint from the previous day. Right now the "saved" session state is an entirely useless feature, it'd be less annoying if it just didn't exist, rather than this constant reminder that it doesn't work correctly.

The only time the save session state works is if I manually close VCS and re-open it, which is just something I don't do. I open it when I start work, then I shut my PC off at the end of the day. I'll have a multitude of windows open, so I don't manually close everything and I shouldn't need to either.

**Comment by wmertens** [comment]
I'll add that I was able to reproduce this on macOS: when my system crashes or when vscode is force-killed, it restores the previous state, not the latest state.

**Comment by youurayy** [comment]
Let me add something to this -- even with a File->Exit (or Cmd+Q) exit, the latest VSCode is not remembering the width of the Side Bar per window.

The whole UI state should be saved. Per window. After any kind of exit/crash.

**Comment by Noiredd** [comment]
It's been 3 years since you were first alarmed of this problem. **Nothing** has been done to fix it, we still keep losing our sessions, and this is still marked as `feature-request`!? At least admit that it's a bug...

**Comment by youurayy** [comment]
Imagine opening 3 or 4 new windows, searching for info and opening all the files you need for your task, carefully aligning ale the open editor tabs to compare data and configs.

Then your laptop crashes in sleep.

You reopen all apps and they restore their UI state without any issues.

You reopen the VSCode and you are back at zero. You have to repeat all your work from the day before, before some people here think that this is an addition feature, and not the most serious bug of the product.

@satyanadella help !

**Comment by chiboreache** [comment]
This is really annoying as fhell !

**Comment by umdstu** [comment]
I'm not sure why people are still complaining about this. It's clear the current team is incapable of this level of development, it's just too difficult for them. Otherwise it would have been done by now. We'll have to wait until they hire, or someone volunteers, one that is competent enough to resolve this aggravating issue. Just don't hold your breath, it may be another 3 years or they're just waiting for Microsoft's future replacement to vscode.

Let's just be glad our browsers have restore tab functionality.

**Comment by youurayy** [comment]
The saving of the layout currently does happen when the editor is quit through File->Exit.
The cheapest way would be to add a 1 minute timer that would be calling this same layout-saving code.
Ideally this would be done by someone familiar with the codebase enough to know if that can collide with extensions, etc.
But in principle it should be really easy.
The problem is that this Issue is buried under tons and tons of other issues. And I'm afraid the PR, if submitted by one of us outsiders, might just get buried under other PR's too.

**Comment by wmertens** [comment]
@umdstu I don't understand how you can complain like that about free software. If it is really so easy to fix for competent programmers, please make the fix, since you're commenting from the viewpoint of a competent programmer.

There is nothing to be gained from commenting in this manner, all you're doing is adding negativity, making sure that this issue is sorted lower on the list.

@youurayy I think a problem is that you can't prevent Electron from being killed until it saved the state. Perhaps the state should be auto-saved every time it changes; not sure why this is not already the case.

**Comment by umdstu** [comment]
@wmertens without complains 'like that', nothing may ever happen. Maybe you're content having this bug for 3 years, I'm not. I also never said it was easy, nor did I say I was competent enough to fix it. If I was, I would have. Therefore it only stands to reason that they dont have those people. That's just logic. Don't presume that I was commenting from any viewpoint other than the statement made.

**Comment by umdstu** [comment]
Also for what it's worth, I highly doubt my 'negativity' is going to make this any lower on the proverbial list than it already is. It's a nice sentiment though.

**Comment by akontsevich** [comment]
> The saving of the layout currently does happen when the editor is quit through File->Exit.
> The cheapest way would be to add a 1 minute timer that would be calling this same layout-saving code.

Problem could be deeper: seems problem happen not because VS Code does not save the state, but it loses, cleans up or even corrupts the state for some reason. So think only project programmers can fix.

**Comment by chiboreache** [comment]
Just making `Exit` available for binding - would be nice workaround, since `Super+Q` won't work

**Comment by mar-ses** [comment]
Just to add my two cents on this: I agree partly with @wmertens, this is a free project, it isn't fair to complain that someone else isn't doing the work necessary to fix an issue.

**However**, I think one of the reasons people are complaining isn't the fact that it's not being fixed, but that it's being swept under the rug. It doesn't seem to feature as a problem that the developers find important, it's not in any iteration plans, and it's just being ignored. It's still marked as a feature-request.

In that case, I think it makes sense for users to continually draw attention to it, and to let the developers know that this bug is a problem that impacts the usability of VSCode.

**Comment by felixpenrose** [comment]
Agreed. Some acknowledgement from a relevant party that this is a bug and will have some eyes look at it in a reasonable time frame would be great.

It's a bug I regularly encounter, not a once in a while problem.

**Comment by umdstu** [comment]
Yes, I understand that this is a free project, but this is a Microsoft product, build by and managed by mostly Microsoft employees. This means there is (mostly) a large corporation running the show (for the most part, yes it's technically open source). I encourage you to read the difference between vscode OSS and Microsoft Visual Studio Code the commercial version. I say all of this only because, although it's 'as is' software, there are certain expectations, warranted or not.

But @mar-ses points out the main problem- lack of feedback. "It will be done in 2021" is much better than *silence*.

Finally- my initial comment was meant to be overly antagonizing, and it's had the desired effect; to draw more attention and activity to the issue at hand :)

**Comment by adbakke** [comment]
I just wish this would **at least** be labeled as a bug, not a feature-request, it really is incorrect. I guess it's not a fun new feature and since it's in the feature request tag no one pays any attention to it.

Noting that comments on this case have taken a turn for the worse as far as tone goes.

**Comment by egamma** [comment]
We are currently working on the January iteration plan and this issue will be on it.

**Comment by umdstu** [comment]
That's awesome @egamma !! Sorry that I had to be an a-hole to get some eyes on this, but I guess it worked. There will be many grateful folks out there once this can be tackled.

Will it be addressed as a Feature Request or a Bug?

**Comment by Noiredd** [comment]
> Noting that comments on this case have taken a turn for the worse as far as tone goes.

True, the community could have behaved more professionally about this. But on the other hand, I have some understanding for grief this problem had caused. A minor bug or lack of a desired feature can become annoying in the long run, but sudden losses of work are extremely frustrating immediately.

Fortunately, we'll soon be able to forget about this at all. Thank you team!!!

**Comment by akontsevich** [comment]
> We are currently working on the January iteration plan and this issue will be on it.

Thanks much!

**Comment by bpasero** [comment]
Pushed a change (https://github.com/microsoft/vscode/pull/87158) that will:
* save state every 60s
* save state when the window looses focus

The real fix is for Electron to provide API to detect the system shutdown case (e.g. what https://github.com/electron/electron/pull/15620 suggests). 

These changes will probably not help when VSCode has focus and windows restarts, though in my testing it works when any non-vscode UI is popping up (e.g. some prompt from Windows to apply updates).

**Comment by Noiredd** [comment]
Thank you!

> The real fix is for (...)

(...) for VSCode to save state any time it changes: when opening and closing windows and tabs. This is the most robust solution.  
@bpasero: detecting shutdown is brittle, would only fix half of the problem. Think of a power loss or a hard reset case - you can't "detect" those, so we would still lose state. Please, **please** keep this in mind if you ever come back to reworking this, because if you later change to relying on Electron, you'll revive this issue.

**Comment by bpasero** [comment]
Yeah I was not suggesting to remove the 60s periodic save interval, but I would not want to make this interval any shorter.

Btw: if Windows decides to just kill VSCode we are at the mercy of how robust SQLite is in those cases (the DB we use for all state). The DB only closes on a proper shutdown, but not on a crash or Windows restart. Having the Electron API might allow us to properly close the DB.

**Comment by bpasero** [comment]
This fix is a tradeoff between having to compute and save the UI state every time it changes (keep in mind: any cursor move is a change in UI state) and protecting against crashes. I many cases we are already storing the UI state when it changes but there are a few exceptions as the example above.

Btw: if Windows decides to just kill VSCode we are at the mercy of how robust SQLite is in those cases (the DB we use for all state). The DB only closes on a proper shutdown, but not on a crash or Windows restart. Having the Electron API might allow us to properly close the DB.

**Comment by akontsevich** [comment]
> The real fix is for Electron to provide API to detect the system shutdown case on Windows

@bpasero Please remember: bug exists on all platforms: Linux and Mac as well, not only Windows.

> Btw: if Windows decides to just kill VSCode we are at the mercy of how robust SQLite is in those cases (the DB we use for all state). The DB only closes on a proper shutdown, but not on a crash or Windows restart. Having the Electron API might allow us to properly close the DB.

**Comment by bpasero** [comment]
@akontsevich and my fix is for all platforms, not just windows

> Then close the DB after every writing.

Thanks for the suggestion, but I hope that is not needed, I trust SQLite to be robust specifically because only 1 process is actually writing to it at all times.

**Comment by akontsevich** [comment]
> @akontsevich and my fix is for all platforms, not just windows. Though from my testing I think I could never reproduce it on any platform but Windows.

On Linux happened many times for me as well in KDE.

**Comment by Noiredd** [comment]
> Though from my testing I think I could never reproduce it on any platform but Windows.

[Here's](https://github.com/microsoft/vscode/issues/12058#issuecomment-473206337) how to reproduce on Linux. Do `killall -9 code` at step 5.

**Comment by bpasero** [comment]
@akontsevich you are right, I see it on Linux as well

**Comment by bpasero** [comment]
@Noiredd yes, I was referring to the state loss on restart while VSCode is running, not a forced kill of the application. I think killing VSCode like that or having a native crash is not the typical scenario we optimize for. The OS restart scenario is more typical to me and should now be fine (pending verification).

There are probably a lot more places in VSCode where a forced process termination leaves state behind. To give an example, all of the extensions are not receiving their `deactivate()` call which could mean that extension state is not persisted.

**Comment by wmertens** [comment]
Since the staye only changes when window layout etc changes, would it not
be possible to persist as soon as it changed? Or can that not be detected?

>

**Comment by bpasero** [comment]
@wmertens most of the UI state is persisted when it actually changes but some is not. As I was mentioning earlier, editor state for example changes every time you move the cursor and is not very fine grained currently (it is the entire editor UI state for all editors persisted under 1 key). I can make another pass to see if the layout state can be persisted when it changes.

**Comment by bpasero** [comment]
**Verification:** open a couple of editors and/or set a specific editor layout and then alt+tab to another app and click the OS restart button. Verify that after restart of the OS when you open VSCode, the window opens in that position and size.
